shell/shared/grain.js

```diff
@@ -1148,52 +1312,68 @@ Router.map(function () {
       ];
     },
 
-    data: function() {
-      if (!this.ready || Meteor.loggingIn()) {
-        return;
-      }
-      if (Meteor.userId() && !Session.get("visit-token-" + this.params.token)) {
-        var tokenInfo = TokenInfo.findOne(this.params.token);
-        if (!tokenInfo || !tokenInfo.apiToken) {
-          this.state.set("error", "invalid authorization token");
-        } else {
-          this.state.set("error", undefined);
-          var apiToken = tokenInfo.apiToken;
-          if (!Grains.findOne({_id: apiToken.grainId, userId: Meteor.userId()}) &&
-              !ApiTokens.findOne({userId: apiToken.userId, "owner.user.userId": Meteor.userId()})) {
+    onRun: function () {
+      var token = this.params.token;
+      var path = this.params.path;
+      var query = this.params.query;
+      var hash = this.params.hash;
+
+      Tracker.autorun(function() {
+        var tokenInfo = TokenInfo.findOne({_id: token});
+        if (tokenInfo && tokenInfo.apiToken) {
+          console.log("valid");
+          var grainId = tokenInfo.apiToken.grainId;
+          if (!Grains.findOne({_id: grainId, userId: Meteor.userId()}) &&
+              !ApiTokens.findOne({userId: tokenInfo.userId, "owner.user.userId": Meteor.userId()})) {
             // The user neither owns the grain nor holds any sturdyrefs from this sharer.
             // Therefore, we ask whether they would like to go incognito.
             // TODO(soon): Base this decision on the contents of the Contacts collection.
-            return {interstitial: true, token: this.params.token};
+          }
+          var grains = globalGrains.get();
+          var grainIndex = grainIdToIndex(grains, grainId);
+          if (grainIndex == -1) {
+            var openView = function openView() {
+              var mainContentElement = document.querySelector("body>.main-content");
+              if (mainContentElement) {
+                var grains = globalGrains.get();
+                var grainToOpen = new GrainView(grainId, path, query, hash, token,
+                    mainContentElement);
+                grainToOpen.openSession();
+                grainIndex = grains.push(grainToOpen) - 1;
+                globalGrains.set(grains);
+                makeGrainIdActive(grainId);
+              } else {
+                Meteor.defer(openView);
+              }
+            };
+            openView();
           } else {
-            Session.set("visit-token-" + this.params.token, "redeem");
+            makeGrainIdActive(grainId);
           }
+
+          var currentGrain = getActiveGrain(grains);
+          if (!currentGrain.shouldShowInterstitial() && currentGrain.sessionStatus() == "closed") {
+            currentGrain.openSession();
```

Comment:
I think the right thing here is for `openSession()` to block on the `GrainView._revealIdentity` variable while it waits for the user to click on an interstitial button. Then we'll only ever need to call `openSession()` once. I'm working on a patch that will do this...